File: 40. Combination_Sum_II/solution(improved).java
class Solution {
    
    List<List<Integer>> ans = new ArrayList<List<Integer>>();
    
    public void recur(int[] candidates, int target, ArrayList<Integer> cur_list, int cur_idx){

        if(target == 0){
            ArrayList<Integer> tmp = new ArrayList<Integer>(cur_list);
            ans.add(tmp);
            return;
        }    
        
        for(int i = cur_idx ; i < candidates.length ; i++){
            
            if(i > cur_idx && candidates[i] == candidates[i - 1])
                continue;
            
            if(target - candidates[i] < 0)
                break;
            
            cur_list.add(candidates[i]);
            
            recur(candidates, target - candidates[i], cur_list, i + 1);
            
            cur_list.remove(cur_list.size() - 1);
        }
        
        return;
        
    }
    
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        
        Arrays.sort(candidates);
        
        ArrayList<Integer> start = new ArrayList<Integer>();
        recur(candidates, target, start, 0);
        
        return ans;
    }
}

File: 40. Combination_Sum_II/solution(improved).cpp
class Solution {
public:
    
    vector<vector<int>> ans;
    
    void recur(vector<int>& cur_vec, int target, vector<int>& candidates, int cur_idx){
        
        if(target == 0){
            ans.push_back(cur_vec);
            return;
        }
        
        for(int i = cur_idx ; i < candidates.size() ; i++){
            
            if(i > cur_idx && candidates[i] == candidates[i - 1])
                continue;
            
            if(target - candidates[i] < 0)
                break;
            
            cur_vec.push_back(candidates[i]);
            
            recur(cur_vec, target - candidates[i], candidates, i + 1);
            
            cur_vec.pop_back();
            
        }
        
        return;
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        
        vector<int> cur_vec;
        sort(candidates.begin(), candidates.end());
        
        recur(cur_vec, target, candidates, 0);
        
        return ans;
        
    }
};